File: backend/utils/validators.py
import re
from typing import Tuple


def validate_email(email: str) -> Tuple[bool, str]:
    """Validate email format.
    
    Returns: (is_valid, error_message)
    """
    if not email:
        return False, "Email is required"
    
    email = email.strip()
    
    # Email regex pattern
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if not re.match(pattern, email):
        return False, "Invalid email format"
    
    if len(email) > 255:
        return False, "Email is too long (max 255 characters)"
    
    return True, ""


def validate_password(password: str) -> Tuple[bool, str]:
    """Validate password strength.
    
    Requirements:
    - Minimum 8 characters
    - At least one uppercase letter
    - At least one lowercase letter
    - At least one digit
    - At least one special character
    
    Returns: (is_valid, error_message)
    """
    if not password:
        return False, "Password is required"
    
    errors = []
    
    if len(password) < 8:
        errors.append("at least 8 characters")
    
    if not re.search(r'[A-Z]', password):
        errors.append("at least one uppercase letter")
    
    if not re.search(r'[a-z]', password):
        errors.append("at least one lowercase letter")
    
    if not re.search(r'\d', password):
        errors.append("at least one digit")
    
    if not re.search(r'[!@#$%^&*()_+\-=\[\]{};:\'",.<>?/\\|`~]', password):
        errors.append("at least one special character")
    
    if errors:
        return False, "Password must contain: " + ", ".join(errors)
    
    if len(password) > 128:
        return False, "Password is too long (max 128 characters)"
    
    return True, ""


def validate_username(username: str) -> Tuple[bool, str]:
    """Validate username format.
    
    Requirements:
    - 3-50 characters
    - Only letters, numbers, dots, hyphens, underscores
    
    Returns: (is_valid, error_message)
    """
    if not username:
        return False, "Username is required"
    
    username = username.strip()
    
    if len(username) < 3:
        return False, "Username must be at least 3 characters"
    
    if len(username) > 50:
        return False, "Username must not exceed 50 characters"
    
    pattern = r'^[a-zA-Z0-9._-]+$'
    if not re.match(pattern, username):
        return False, "Username can only contain letters, numbers, dots, hyphens, and underscores"
    
    return True, ""


def validate_role(role: str) -> Tuple[bool, str]:
    """Validate user role.
    
    Returns: (is_valid, error_message)
    """
    if not role:
        return False, "Role is required"
    
    valid_roles = ['Employee', 'HR Manager', 'Payroll Manager', 'Admin']
    if role not in valid_roles:
        return False, f"Invalid role. Must be one of: {', '.join(valid_roles)}"
    
    return True, ""


def validate_login_input(username: str, password: str) -> Tuple[bool, str]:
    """Validate login form inputs.
    
    Returns: (is_valid, error_message)
    """
    if not username:
        return False, "Email/Username is required"
    
    if not password:
        return False, "Password is required"
    
    if len(username) > 255:
        return False, "Email/Username is too long"
    
    if len(password) > 128:
        return False, "Password is too long"
    
    return True, ""

File: frontend/src/utils/validators.js
/**
 * Frontend validation utilities for user authentication
 */

export const validateEmail = (email) => {
  if (!email) return { valid: false, error: 'Email is required' };
  
  const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  if (!pattern.test(email)) {
    return { valid: false, error: 'Invalid email format' };
  }
  
  if (email.length > 255) {
    return { valid: false, error: 'Email is too long (max 255 characters)' };
  }
  
  return { valid: true, error: '' };
};

export const validatePassword = (password) => {
  if (!password) return { valid: false, error: 'Password is required' };
  
  const errors = [];
  
  if (password.length < 8) {
    errors.push('at least 8 characters');
  }
  
  if (!/[A-Z]/.test(password)) {
    errors.push('at least one uppercase letter');
  }
  
  if (!/[a-z]/.test(password)) {
    errors.push('at least one lowercase letter');
  }
  
  if (!/\d/.test(password)) {
    errors.push('at least one digit');
  }
  
  if (!/[!@#$%^&*()_+\-=\[\]{};:'"",.<>?/\\|`~]/.test(password)) {
    errors.push('at least one special character');
  }
  
  if (errors.length > 0) {
    return { valid: false, error: 'Password must contain: ' + errors.join(', ') };
  }
  
  if (password.length > 128) {
    return { valid: false, error: 'Password is too long (max 128 characters)' };
  }
  
  return { valid: true, error: '' };
};

export const validateUsername = (username) => {
  if (!username) return { valid: false, error: 'Username is required' };
  
  username = username.trim();
  
  if (username.length < 3) {
    return { valid: false, error: 'Username must be at least 3 characters' };
  }
  
  if (username.length > 50) {
    return { valid: false, error: 'Username must not exceed 50 characters' };
  }
  
  const pattern = /^[a-zA-Z0-9._-]+$/;
  if (!pattern.test(username)) {
    return { valid: false, error: 'Username can only contain letters, numbers, dots, hyphens, and underscores' };
  }
  
  return { valid: true, error: '' };
};

export const validateLoginForm = (username, password) => {
  if (!username) return { valid: false, error: 'Email/Username is required' };
  if (!password) return { valid: false, error: 'Password is required' };
  
  if (username.length > 255) {
    return { valid: false, error: 'Email/Username is too long' };
  }
  
  if (password.length > 128) {
    return { valid: false, error: 'Password is too long' };
  }
  
  return { valid: true, error: '' };
};

/**
 * Get password strength score (0-4)
 * Used for visual feedback
 */
export const getPasswordStrength = (password) => {
  if (!password) return 0;
  
  let strength = 0;
  
  if (password.length >= 8) strength++;
  if (/[A-Z]/.test(password)) strength++;
  if (/[a-z]/.test(password)) strength++;
  if (/\d/.test(password)) strength++;
  
  return strength;
};

export const getPasswordStrengthLabel = (strength) => {
  switch (strength) {
    case 0:
    case 1:
      return 'Weak';
    case 2:
      return 'Fair';
    case 3:
      return 'Good';
    case 4:
      return 'Strong';
    default:
      return '';
  }
};
